lib/autocomplete-manager.coffee

```diff
@@ -272,18 +180,166 @@ class AutocompleteManager
+      @contentsModified()
+    else
+      @hideSuggestionList()
+
+  #  |||              |||
+  #  vvv PROVIDER API vvv
+
+  registerProviderForGrammars: (provider, grammars) =>
+    return unless provider?
+    return unless grammars? and _.size(grammars) > 0
+    grammars = _.filter(grammars, (grammar) -> grammar?.scopeName?)
+    scopes = _.pluck(grammars, 'scopeName')
+    return @registerProviderForScopes(provider, scopes)
+
+  registerProviderForScopes: (provider, scopes) =>
+    return unless provider?
+    return unless scopes? and _.size(scopes) > 0
+    for scope in scopes
+      existing = _.findWhere(_.keys(@scopes), scope)
+      if existing? and @scopes[scope]?
+        @scopes[scope].push(provider)
+        @scopes[scope] = _.uniq(@scopes[scope])
+      else
+        @scopes[scope] = [provider]
+
+    if provider.dispose?
+      @subscriptions.add(provider) unless _.contains(@subscriptions, provider)
+
+    new Disposable =>
+      @unregisterProviderForScopes(provider, scopes)
+
+  registerProviderForEditor: (provider, editor) =>
+    return unless provider?
+    return unless editor?
+    grammar = editor?.getGrammar()
+    return unless grammar?
+    return if grammar.scopeName is 'text.plain.null-grammar'
+    return @registerProviderForGrammars(provider, [grammar])
+
+  unregisterProviderForGrammars: (provider, grammars) =>
+    return unless provider?
+    return unless grammars? and _.size(grammars) > 0
+    grammars = _.filter(grammars, (grammar) -> grammar?.scopeName?)
+    scopes = _.pluck(grammars, 'scopeName')
+    return @unregisterProviderForScopes(provider, scopes)
+
+  unregisterProviderForScopes: (provider, scopes) =>
+    return unless provider?
+    return unless scopes? and _.size(scopes) > 0
+
+    for scope in scopes
+      existing = _.findWhere(_.keys(@scopes), scope)
+      if existing?
+        @scopes[scope] = _.filter(@scopes[scope], (p) -> p isnt provider)
+        delete @scopes[scope] unless _.size(@scopes[scope]) > 0
+
+    # @subscriptions.remove(provider) unless @providerIsRegistered(provider)
+
+  providerIsRegistered: (provider, scopes) =>
+    # TODO
+    return true
+
+  unregisterProviderForEditor: (provider, editor) =>
+    return unless provider?
+    return unless editor?
+    grammar = editor?.getGrammar()
+    return unless grammar?
+    return @unregisterProviderForGrammars(provider, [grammar])
+
+  unregisterProvider: (provider) =>
+    return unless provider?
+    return @unregisterProviderForScopes(provider, _.keys(@scopes))
+    @subscriptions.remove(provider) if provider.dispose?
+
+  provideApi: =>
+    atom.services.provide 'autocomplete.provider-api', "1.0.0", { @registerProviderForEditor, @unregisterProviderForEditor, @unregisterProvider, Provider, Suggestion }
+    atom.services.provide 'autocomplete.provider-api', '2.0.0', { @registerProviderForGrammars, @registerProviderForScopes, @unregisterProviderForGrammars, @unregisterProviderForScopes, @unregisterProvider }
```

Comment:
The only reason this might be difficult is if we had a provider that wanted to be able to provide configuration for what scopes or grammars to which it is applied.  However, in that case it could just be disposed and reregistered when the configuration changes.